docs/adr/0007-modelling-history.md

```diff
@@ -0,0 +1,272 @@
+  published documents)
+
+A [MoSCoW](https://en.wikipedia.org/wiki/MoSCoW_method) breakdown of this
+functionality was defined as:
+
+### Must
+
+- Store full information on latest and live versions of a document
+- Consider how a revision of a document and aspects of it have state that may
+  change
+
+### Should
+
+- Have the full information for every edition that was published on GOV.UK
+- Record every change a publisher makes and have that information accessible
+- Consider states to be something that may change in a background job
+- Avoid relying on the TimelineEntry model as the sole answer to any historical
+  queries
+
+### Could
+
+- Link TimelineEntry to changes rather than storing the information twice
+- Enable a future feature to enable reverting to different points in time
+- Separate related data to different tables to reduce data usage
+
+### Won’t
+
+- Consider the sharing of data between translations, as per
+  [ADR-3](../adr/0003-initial-domain-modelling.md#decision)
+
+## Decision
+
+The proposed data model expands on the definitions of entities accordingly:
+
+**Document**: A record that represents all versions of a piece of content in a
+particular locale. It has many editions and many revisions. At any time it will
+have a current edition -  shown on Content Publisher index - and potentially a
+live edition which is currently on GOV.UK. The live and current edition can be
+the same. Document serves as a mutable entity that tracks when a Document was
+last edited and it is expected to be a joining point for document related data
+that is not associated with a particular editions.
+
+**Edition**: An edition is a numbered version of a document that has been, or is
+expected to be, published on GOV.UK. It is associated with a current revision
+and maintains an association with all revisions that have been associated with
+it. It has a current state (e.g. draft, submitted, published)
+and has knowledge of the past states it held. It is mutable and is expected
+as a place where any edition level database constraints can be placed.
+
+**Revision**: Represents the content of a document at a particular point in
+time. It has a number to indicate which revision of the document it is and
+stores who created it. Any time a user changes an aspect of a document a new
+revision is created and the current edition is updated to reference that
+revision - a revision cannot be edited and is considered immutable. A revision
+may have a association with a collection of items, such as images, when a new
+revision is created new images are not created but a new join between them is.
+
+This modelling is represented in the following diagram:
+
+![Class diagram](0007/class-diagram.png)
+
+Note: for brevity this diagram shows a reduced selection of fields and removes
+associations with Users.
+
+The purpose of models that not described above are:
+
+**State**: This represents the state of a document at a particular point in time,
```

Comment:
> Could you explain why tracking asynchronous changes implies making the state mutable?

Does this section https://github.com/alphagov/content-publisher/blob/adr-7/docs/adr/0007-modelling-history.md#edition-state-changed-asynchronously help at all? If not I can have another attempt at phrasing but initially I might just paraphrase that section.

A State model should represent a single action by a user (or scheduled task in case of scheduling) to change the state of an edition. In an asynchronous world that state change will result in some form of Publishing API syncing, the mutability aspect of that is to track that.

> And another reasoning question: why would we want to ask the state about the reason for retiring? I'm thinking maybe we're showing it on the document summary or something, but that seems like a specific reason and would be good to understand the general idea behind the association to specifics.

The logic behind that association is that the reason for retiring is a property of the retired state. If the edition is no longer in a retired state then that reason is no longer a property of a document. So user input that only has a relevance in a particular state (and presumably taken at the time of the state change) should be associated with the state.

> Could we consider prefixing this model with something, as it's quite a generic term that's also ambiguous between 'the state we store about something' and 'metadata about something, somewhere'. Perhaps WorkflowState or PublishingState?

Yup, I'd been thinking it stood out unprefixed. I can't really think of anything than EditionState or EditionStatus though as it's primary role is storing the info we describe in the UI as status.